chore: use `package.json` directory to resolve output directory

aslushnikov opened the pull request:
This patch:
- starts using directory of `package.json` to resolve default
  output directory path
- starts using either `package.json` directory or configuration
  directory to resolve all relative paths

References #12970

yury-s commented on packages/playwright-test/src/loader.ts:
```diff
@@ -37,7 +37,8 @@ export class Loader {
   private _fullConfig: FullConfig;
-  private _config: Config = {};
+  private _configDir: string = '';
+  private _packageJsonDir: string | undefined;
```
drop it as it's unused

dgozman reviewed (commented):
- [x] As discussed, let's undo the rootDir change.
- [x] Change docs to mention this.
- [x] Let's add a test for the new behavior.
- [ ] Might as well put html report change here.

aslushnikov committed:
feat: use `package.json` directory as a default for `outputDir`

This patch starts using `package.json` directory as a default
directory for `outputDir`.

Drive-by: refactore config parsing logic.
